module/zfs/vdev_draid.c

```diff
@@ -0,0 +1,1661 @@
+		rc->rc_abd = abd_alloc_linear(rc->rc_size, B_TRUE);
+	}
+
+	/*
+	 * Map buffers for data columns and allocate/map buffers for skip
+	 * sectors.  There are three distinct cases for dRAID which are
+	 * required to support sequential rebuild.
+	 */
+	if (zio->io_type == ZIO_TYPE_WRITE) {
+		vdev_draid_map_alloc_write(zio, rm);
+	} else if ((rm->rm_nskip > 0) &&
+	    (zio->io_flags & (ZIO_FLAG_SCRUB | ZIO_FLAG_RESILVER))) {
+		vdev_draid_map_alloc_scrub(zio, rm);
+	} else {
+		ASSERT3U(zio->io_type, ==, ZIO_TYPE_READ);
+		vdev_draid_map_alloc_read(zio, rm);
+	}
+
+	rm->rm_ops = vdev_raidz_math_get_ops();
+	zio->io_vsd = rm;
+	zio->io_vsd_ops = &vdev_raidz_vsd_ops;
+
+	return (rm);
+}
+
+/*
+ * Converts a dRAID read raidz_map_t to a dRAID scrub raidz_map_t.  The
+ * key difference is that an ABD is allocated to back up skip sectors
+ * so they may be read, verified, and repaired if needed.
+ */
+void
+vdev_draid_map_include_skip_sectors(zio_t *zio)
+{
+	raidz_map_t *rm = zio->io_vsd;
+
+	ASSERT3U(zio->io_type, ==, ZIO_TYPE_READ);
+	ASSERT3P(rm->rm_abd_skip, ==, NULL);
+
+	for (uint64_t c = rm->rm_firstdatacol; c < rm->rm_cols; c++) {
+		ASSERT(!abd_is_gang(rm->rm_col[c].rc_abd));
+		abd_put(rm->rm_col[c].rc_abd);
+	}
+
+	vdev_draid_map_alloc_scrub(zio, rm);
+}
+
+/*
+ * Converts a logical offset to the corresponding group number.
+ */
+uint64_t
+vdev_draid_offset_to_group(vdev_t *vd, uint64_t offset)
+{
+	vdev_draid_config_t *vdc = vd->vdev_tsd;
+
+	ASSERT3P(vd->vdev_ops, ==, &vdev_draid_ops);
+
+	return (offset / vdc->vdc_groupsz);
+}
+
+/*
+ * Converts a group number to the logical starting offset for that group.
+ */
+uint64_t
+vdev_draid_group_to_offset(vdev_t *vd, uint64_t group)
+{
+	vdev_draid_config_t *vdc = vd->vdev_tsd;
+
+	ASSERT3P(vd->vdev_ops, ==, &vdev_draid_ops);
+
+	return (group * vdc->vdc_groupsz);
+}
+
+/*
+ * Given a offset into a draid, compute a group aligned offset.
+ */
+uint64_t
+vdev_draid_get_astart(vdev_t *vd, const uint64_t start)
+{
+	vdev_draid_config_t *vdc = vd->vdev_tsd;
+
+	ASSERT3P(vd->vdev_ops, ==, &vdev_draid_ops);
+
+	return (roundup(start, vdc->vdc_groupwidth << vd->vdev_ashift));
+}
+
+/*
+ * Return the asize which is the psize rounded up to a full group width.
+ * i.e. vdev_draid_psize_to_asize().
+ */
+static uint64_t
+vdev_draid_asize(vdev_t *vd, uint64_t psize)
+{
+	vdev_draid_config_t *vdc = vd->vdev_tsd;
+	uint64_t ashift = vd->vdev_ashift;
+	uint64_t asize = ((psize - 1) >> ashift) + 1;
+
+	ASSERT3P(vd->vdev_ops, ==, &vdev_draid_ops);
+
+	asize = roundup(asize, vdc->vdc_ndata);
+	asize += vdc->vdc_nparity * (asize / vdc->vdc_ndata);
+
+	ASSERT3U(asize, !=, 0);
+	ASSERT3U(asize, <, vdc->vdc_groupsz);
+	ASSERT3U(asize % (vdc->vdc_groupwidth), ==, 0);
+
+	return (asize << ashift);
+}
+
+/*
+ * Deflate the asize to the psize for a block at the provided offset.
+ * This involves stripping parity.
+ */
+uint64_t
+vdev_draid_asize_to_psize(vdev_t *vd, uint64_t asize, uint64_t offset)
```

Comment:
I don't see where `offset` is relevant.

Reply:
Good catch.  This was leftover from before the tiling changes.  Removed.